libkbfs/mdserver_tlf_storage.go

```diff
@@ -0,0 +1,596 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"golang.org/x/net/context"
+
+	keybase1 "github.com/keybase/client/go/protocol"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type mdServerTlfStorage struct {
+	codec  Codec
+	clock  Clock
+	crypto cryptoPure
+	dir    string
+
+	// Protects any IO operations in dir or any of its children,
+	// as well all the DBs, truncateLockHolder, and isShutdown.
+	//
+	// TODO: Consider using https://github.com/pkg/singlefile
+	// instead.
+	lock sync.RWMutex
+
+	// TODO: Replace the DBs below with a journal.
+
+	mdDb     *leveldb.DB // [branchId]+[revision] -> MdID
+	branchDb *leveldb.DB // deviceKID             -> branchId
+
+	// Store the truncate lock holder just in memory, so it gets
+	// wiped after a restart.
+	truncateLockHolder *keybase1.KID
```

Comment:
Just a heads up -- we'll want to plumb truncate calls through to the real MD server (or redirect them at a higher level) even when we're in async write mode.    I guess we'll still want support for it here for tests though.  No suggestions on things to change now, just mentioning it for the future.